mindmeld/components/entity_resolver.py

```diff
@@ -562,38 +1135,597 @@ def _predict_exact_match(self, entity):
+                                in synonyms_encodings
+            synonyms_encodings (np.array): a 2d array of embedding of the synonyms; an array of
+                                            size equal to number of synonyms
+            entity_emb (np.array): a 2d array of embedding of the input entity text(s)
+            top_n (int): maximum number of results to populate
+        Returns:
+            Union[dict, list[tuple]]: if return_as_dict, returns a dictionary of synonyms and their
+                                        scores, else a list of sorted synonym names, paired with
+                                        their similarity scores (descending)
+        """
+
+        n_entities = len(entity_emb)
+
+        is_single = n_entities == 1
+
+        # [n_syns, emd_dim] -> [n_entities, n_syns, emd_dim]
+        t_syn_enc = _torch("as_tensor", synonyms_encodings)
+        t_syn_enc = t_syn_enc.expand([n_entities, *t_syn_enc.shape])
+
+        # [n_entities, emd_dim] -> [n_entities, n_syns, emd_dim]
+        t_entity_emb = _torch("as_tensor", entity_emb)
+        t_entity_emb = t_entity_emb.unsqueeze(dim=1).expand_as(t_syn_enc)
+
+        # returns -> [n_entities, n_syns]
+        similarity_scores_2d = _torch(
+            "cosine_similarity", t_syn_enc, t_entity_emb, dim=-1).numpy()
+
+        results = []
+        for similarity_scores in similarity_scores_2d:
+            similarity_scores = similarity_scores.reshape(-1)
+            similarity_scores = np.around(similarity_scores, decimals=2)
+
+            if return_as_dict:
+                results.append(dict(zip(synonyms.keys(), similarity_scores)))
+            else:
+                # results in descending scores
+                n_scores = len(similarity_scores)
+                if n_scores > top_n:
+                    top_inds = similarity_scores.argpartition(n_scores - top_n)[-top_n:]
+                    result = sorted(
+                        zip(np.asarray([*synonyms.keys()])[top_inds], similarity_scores[top_inds]),
+                        key=lambda x: x[1], reverse=True)
+                else:
+                    result = sorted(zip(synonyms.keys(), similarity_scores), key=lambda x: x[1],
+                                    reverse=True)
+                results.append(result)
+
+        if is_single:
+            return results[0]
+
+        return results
+
+    def _fit(self, clean, entity_map):
+
+        if clean and os.path.exists(self.cache_path):
+            os.remove(self.cache_path)
+
+        # load mapping.json data and process it
+        augment_lower_case = self._runtime_configs["augment_lower_case"]
+        self._processed_entity_map = self._process_entity_map(
+            self.type,
+            entity_map,
+            augment_lower_case=augment_lower_case
+        )
+
+        # load embeddings from cache if exists, encode any other synonyms if required
+        synonyms, synonyms_embs = OrderedDict(), np.empty(0)
```

Comment:
Using a dictionary instead of a index mapping dict and a numpy array is faster. See below testing script:
```python
import pickle
from time import time

import numpy as np

n_times = 1000

data = {str(i): np.ones(768, dtype=float) for i in range(1000, 2500)}
keys = {key: j for j, key in enumerate([*data.keys()])}
print(len(keys))
values = np.array([*data.values()])
print(values.shape)

st_time = time()
for _ in range(n_times):
    opfile = open("temp_embeddings.pkl", "wb")
    pickle.dump(data, opfile)
    opfile.close()
    opfile = open("temp_embeddings.pkl", "rb")
    pickle.load(opfile)
    opfile.close()
print("time: {:.2f} ms per save-load loop".format(1000*(time() - st_time)/n_times))

st_time = time()
for _ in range(n_times):
    data = {"keys": keys, "values": values}
    opfile = open("temp_embeddings.pkl", "wb")
    pickle.dump(data, opfile)
    opfile.close()
    opfile = open("temp_embeddings.pkl", "rb")
    pickle.load(opfile)
    opfile.close()
print("time: {:.2f} ms per save-load loop".format(1000*(time() - st_time)/n_times))

```

This leads to the result:
```bash
1500
(1500, 768)
time: 27.37 ms per save-load loop
time: 10.53 ms per save-load loop
```